flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/BoundedProcessingOverRowProcessFunction.scala

```diff
@@ -0,0 +1,182 @@
+      new ListTypeInfo[Row](inputType).asInstanceOf[TypeInformation[JList[Row]]]
+    
+    val mapStateDescriptor: MapStateDescriptor[Long, JList[Row]] =
+      new MapStateDescriptor[Long, JList[Row]]("windowBufferMapState",
+        BasicTypeInfo.LONG_TYPE_INFO.asInstanceOf[TypeInformation[Long]], rowListTypeInfo)
+    
+    rowMapState = getRuntimeContext.getMapState(mapStateDescriptor)
+
+    val stateDescriptor: ValueStateDescriptor[Row] =
+      new ValueStateDescriptor[Row]("aggregationState", aggregatesTypeInfo)
+      
+    accumulatorState = getRuntimeContext.getState(stateDescriptor)
+    
+    val processedCountDescriptor : ValueStateDescriptor[Long] =
+       new ValueStateDescriptor[Long]("processedCountState", classOf[Long])
+    
+    counterState = getRuntimeContext.getState(processedCountDescriptor)
+    
+    val smallesTimestampDescriptor : ValueStateDescriptor[Long] =
+       new ValueStateDescriptor[Long]("smallesTSState", classOf[Long])
+    
+    smallestTsState = getRuntimeContext.getState(smallesTimestampDescriptor)
+    
+  }
+
+  override def processElement(
+    input: Row,
+    ctx: ProcessFunction[Row, Row]#Context,
+    out: Collector[Row]): Unit = {
+    
+    val currentTime = ctx.timerService().currentProcessingTime()
+    var i = 0
+
+    accumulators = accumulatorState.value()
+    // initialize state for the first processed element
+    if(accumulators == null){
+      accumulators = new Row(aggregates.length)
+      while (i < aggregates.length) {
+        accumulators.setField(i, aggregates(i).createAccumulator())
+        i += 1
+      }
+    }
+    
+    // get smallest timestamp 
+    smallestTs = smallestTsState.value()
+    if(smallestTs == 0L){
+      smallestTs = currentTime
+    }
+    // get previous counter value
+    counter = counterState.value()
+    
+    if (counter == precedingOffset) {
+      val retractTs = smallestTs
+      val retractList = rowMapState.get(smallestTs)
+
+      // get oldest element beyond buffer size   
+      // and if oldest element exist, retract value
+      i = 0
+      while (i < aggregates.length) {
+        val accumulator = accumulators.getField(i).asInstanceOf[Accumulator]
+        aggregates(i).retract(accumulator, retractList.get(0).getField(aggFields(i)))
+        i += 1
+      }
+      retractList.remove(0)
+      counter -= 1
+      // if reference timestamp list not empty, keep the list
+      if (!retractList.isEmpty()) {
+        rowMapState.put(retractTs, retractList)
+      } // if smallest timestamp list is empty, remove and find new smallest
+      else {
+        rowMapState.remove(retractTs)
+        val iter = rowMapState.keys.iterator()
+        var currentTs: Long = 0L
+        var newSmallesTs: Long = Long.MaxValue
+        while(iter.hasNext){
+          currentTs = iter.next
+          if(currentTs < newSmallesTs){
+            newSmallesTs = currentTs
+          }
+        }
+        smallestTs = newSmallesTs
```

Comment:
We can immediately update the `smallestTsState` here and avoid updating if if the timestamp was not changed.